### PR TITLE
interfaces: auto-connect interface required by livepatch

### PR DESCRIPTION
**Pull request by arges**
This allows an snap app given a particular name/developer combination to
auto-connect interfaces that are not autoConnect and may be restricted.
In addition tests are added to ensure that name/developer combinations not
in this list can't auto connect interfaces.

### CODE REVIEW

**Comment by zyga** [comment]
I'm +1/-1 on this.

I think we should either allow all of "canonical", as a special one-off case, or list individual snaps only (and in that case, the publisher name is meaningless sine without assertions it is easy to spoof).

File: interfaces/repo.go [hunk 1]
```diff
@@ -726,6 +726,17 @@ func (r *Repository) DisconnectSnap(snapName string) ([]string, error) {
 	return result, nil
 }
 
+// isSpecialAutoConnectSnap checks special Name/Developer combinations to see
+// if this particular snap's connections should be automatically connected even
+// if the interfaces are not autoconnect and the snap is not an OS snap.
+// TODO: When assertions are in place this code should be removed.
+func isSpecialAutoConnectSnap(snap *snap.Info) bool {
+	if snap.Name() == "canonical-livepatch" && snap.Developer == "canonical" {
```
Comment (chipaca) [inline interfaces/repo.go]: and `snap.Revision.Store()`?

Comment (arges) [inline interfaces/repo.go]: I can add that, but what does it do exactly?

Comment (niemeyer) [inline interfaces/repo.go]: @chipaca I believe the check for "canonical" already sorts that out. We only get the developer filed when it comes from the store, right?

File: interfaces/repo.go [hunk 2]
```diff
@@ -736,9 +747,11 @@ func (r *Repository) AutoConnectCandidates(plugSnapName, plugName string) []*Slo
 	if plug == nil {
 		return nil
 	}
-	if r.ifaces[plug.Interface].AutoConnect() == false {
+
+	if (r.ifaces[plug.Interface].AutoConnect() == false) && !(isSpecialAutoConnectSnap(plug.Snap)) {
```
Comment (niemeyer) [inline interfaces/repo.go]: It's a bit awkward to do that at this level. It's ignoring what the interface is entirely.

I'll complement this comment below.

File: interfaces/repo.go [hunk 3]
```diff
@@ -736,9 +747,11 @@ func (r *Repository) AutoConnectCandidates(plugSnapName, plugName string) []*Slo
...
 		return nil
 	}
+
 	var candidates []*Slot
 	for _, slotsForSnap := range r.slots {
 		for _, slot := range slotsForSnap {
```
Comment (niemeyer) [inline interfaces/repo.go]: The logic below is also more special casing. This is getting a bit wild, and that's likely a side effect of a misdesign in our interface API. I believe we can easily sort all of these special cases out by having this instead:
```
iface := r.ifaces[plug.Interface]
...
        if iface.AutoConnect(plug, slot) {
                candidates = append(candidates, slot)
        }
```
We then move the custom logic to inside the respective interfaces.

Comment (arges) [inline interfaces/repo.go]: ACK. This makes sense. I'll work on a 2 part patch here. 1 to refactor to AutoConnect(plug, slot) and the other for custom logic for my interface.

Comment (arges) [inline interfaces/repo.go]: This seems like a fairly invasive change, is this also going to be useful when Assertions are working, etc?

**Comment by arges** [comment]
In response to:
https://github.com/snapcore/snapd/pull/1815#discussion_r77182658

I'm posting this work in progress to get some early feedback since this is turning into a much more massive change.

So by modifying AutoConnect to take in plug, slot and handle logic here we have a lot of functions / tests that get changed.

In addition I still need to move the content tests from interfaces/repo_test.go into interfaces/builtin/content_test.go. I'm not sure how easy it will be to do a mockup repo at that level, or if it makes sense to keep the unit tests at the repo level.

Another challenge is modifying the commonInterface type builtin interfaces to use an AutoConnect function instead of a boolean variable. Here for any necessary interfaces that need to implement this I need re-write these to not use commonInterface (which for me is most of the interfaces I want). I tried to hack commonInterface so I can override AutoConnect but that seems really hackish.

So at this point any guidance, this is looking like a massive change so either there is a better way to do this, or I'm missing something.

--chris

**Pull request closed without merging**